### ISSUE TITLE
Burly with window(s) containing shell 

### ISSUE THREAD

**Issue by jacquesdurden** [alphapapa/burly.el#22]
Burly is great! Thank you for this super handy tool. 

I have only a little problem, or maybe it's actually a new feature request. 

I'd like to save a layout of say 3 windows in my frame with one of the window open on a shell (using just ```shell```). My wish would be that by invoking the Burly bookmark of this layout, the 3 windows would reappear with my shell reopened exactly as before. 

For now, while Burly is perfectly fine for standard windows open on files, I got this error message if one of them contains a shell ```Wrong type argument: stringp, nil``` 

Thank you.

**Comment by alphapapa** [comment]
Hi Jacques,

Thanks, I'm glad you like it.

Supporting the various kinds of shell/terminal buffers would best be handled by giving each of those buffers' major modes a `bookmark-make-record-function` so their buffers could be bookmarked by Emacs.  Ideally those functions would be merged into Emacs upstream so that everyone would benefit from them.

Other than that, we could add some mode-specific workarounds in this package, but we'd have to write essentially the same code, so it might as well be upstreamed.

I rarely keep shells open in Emacs, so I probably wouldn't use this feature very much right now.  Would you be able/willing to work on that?

> For now, while Burly is perfectly fine for standard windows open on files, I got this error message if one of them contains a shell Wrong type argument: stringp, nil

That part is a bug, so thanks for reporting that.

**Comment by captain-yoshi** [comment]
This module is great (elegant, simple and effective). Like @jacquesdurden I would like to store eshell with Burly.

I have come across this [post](https://fuco1.github.io/2017-10-08-Using-bookmarks-with-eshell-and-docker-tramp.html) that led me to this eshell bookmark [github repo](https://github.com/Fuco1/eshell-bookmark). I didn't have the chance to test this, but this could be a good starting point and/or solution.

**Comment by jacquesdurden** [comment]
Thank you for this solution @captain-yoshi! I tried it and it works! (just be sure though to well install ```eshell-bookmark``` with its activation command in your init.el, see its github page) This could be a good info for the Burly team too. I just have really minor positioning issues in the window containing eshell at startup but I can live with this. Thank you again.

**Comment by captain-yoshi** [comment]
Glad it could help! I'm currently using doom emacs and have problem opening the bookmark...
```elisp
Warning (burly): Error while opening bookmark: ERROR:(void-function eshell-bookmark--restore)  RECORD:("" (defaults "*doom:eshell*") (filename . "~/Downloads/") (handler . eshell-bookmark--restore))
```

config.el
```elisp
(add-hook! 'eshell-mode-hook #'eshell-bookmark-setup)
```
packages.el
```elisp
(package! eshell-bookmark)
```

@jacquesdurden Is my configuration good? I'm fairly new to elisp and emacs. Thx

**Comment by jacquesdurden** [comment]
I don't use doom emacs but why have you "!" and "#" in you commands? It seems you must just write 

``
(add-hook 'eshell-mode-hook 'eshell-bookmark-setup)
``

Have you also tried with 

``
(use-package eshell-bookmark
  :after eshell
  :config
  (add-hook 'eshell-mode-hook #'eshell-bookmark-setup))
``

instead?

**Comment by captain-yoshi** [comment]
I think doom emacs uses the "!" and "#", see the [documentation](https://github.com/hlissner/doom-emacs/blob/develop/docs/getting_started.org#configuring-packages). Maybe @alphapapa can help with burly's warning.

> (add-hook 'eshell-mode-hook 'eshell-bookmark-setup)

Gives me the same warning as my previous post.

**Comment by alphapapa** [comment]
@captain-yoshi Thanks for sharing the links.  I added @Fuco1's article and package to the notes file.  Ideally the eshell-bookmark package would be upstreamed into Emacs.  Since I don't much use eshell (yet?), I'll leave that to you to pursue.

Regarding that error: it means what it says: the function is void.  That means you haven't loaded it into your Emacs.

Since Doom and generic Emacs issues are off-topic here, please use an appropriate forum to seek help with them.

Thanks.

**Comment by Fuco1** [comment]
I agree it would be nice to push upstream, I think someone even tried to do it already.  Unfortunately I'm quite unable to tend to my packages these days, but it's certainly open for anyone (I have signed the fsf papers).

**Comment by alphapapa** [comment]
AFAIK there is no longer a bug here with regard to Burly.

The enhancement to `shell` in Emacs itself, to add bookmark support, is still needed.  Maybe someone will get to that someday...
